### PR TITLE
Add cpumanager policy options

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:
Split from the parent PR https://github.com/kubernetes/kubernetes/pull/101432 . The parent PR want to introduce a new commandline flag for kubelet, and we isolate the API-changing commits here. Please check https://github.com/kubernetes/kubernetes/pull/101432 for the full context.

#### Which issue(s) this PR fixes:
N/A

#### Special notes for your reviewer:
The split was  suggested in https://github.com/kubernetes/kubernetes/pull/101432#issuecomment-831873880 to facilitate the reviews and to accelerate the process.

#### Does this PR introduce a user-facing change?
```release-note
Add the --cpumanager--policy-options to support fine tuning of the cpumanager static policy
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
- KEP: https://github.com/kubernetes/enhancements/issues/2625
- Usage: https://github.com/kubernetes/kubernetes/pull/101432

### CODE REVIEW

**Comment by fromanirh** [comment]
@thockin howdy! just checking if there is something I can help with for the API review!

**Comment by thockin** [comment]
I'm assigned for API review and while this is TECHNICALLY an API, it's much more specifically node-owned.  I did not get to review the KEP, but I have one pretty basic question.  Do we really want this as a `[]string` or do we want to decode to a `map[string]string` or similar?  E.g. the example "reject-non-smt-aligned" is an implied boolean - will we ever need non-boolean params?  History tells me that these things tend to expand.  

Of course we could say that "foo=bar" is a single string and the consumer must decode it.  Is that what we are saying?

**Comment by fromanirh** [comment]
> I'm assigned for API review and while this is TECHNICALLY an API, it's much more specifically node-owned. 

Thanks for chiming in! I'm not sure what's the best approach here (for the reasons you just outlined) so I'm sticking to the book

> I did not get to review the KEP, but I have one pretty basic question. Do we really want this as a `[]string` or do we want to decode to a `map[string]string` or similar? E.g. the example "reject-non-smt-aligned" is an implied boolean - will we ever need non-boolean params? History tells me that these things tend to expand.
> 
> Of course we could say that "foo=bar" is a single string and the consumer must decode it. Is that what we are saying?

Atm yes. The concrete use cases I know for sure are
- https://github.com/kubernetes/enhancements/pull/2626 (accepted for 1.22)
-  https://github.com/kubernetes/enhancements/pull/2684 (possibly, planned for 1.23)

So we know for sure that we need implied boolean. Now, @klueska mentioned (also above) use cases with key=value, and it seems fair to me to extend in this direction and decode cpumanager options in `map[string]string` like you suggested.

I think this is the sweet spot which best serves all the cases we have on the radar, and should be reasonnable forward looking. I don't think atm we should extend further than that.

So next step is: I can update this PR and rebase https://github.com/kubernetes/kubernetes/pull/101432 right after, so we can move forward. Works for everyone?

**Comment by klueska** [comment]
@thockin
Although, this is mostly a kubelet-only change, it still requires approval from someone higher up the hierarchy because of the changes to:
```
pkg/features/kube_features.go
staging/src/k8s.io/kubelet/*
```

I am an owner in `pkg/kubelet` and am doing the primary review for this work (as implemented in https://github.com/kubernetes/kubernetes/pull/101432), but I don't have the ability to approve changes to the files listed above.

I had @fromanirh split out these API level changes to this PR so that they could more easily be reviewed by you (or someone else that typically does API reviews) so we can get this API finalized and merged before focusing on the meat of the implementation (which I will then have full power to approve).

**Comment by ehashman** [comment]
> Although, this is mostly a kubelet-only change, it still requires approval from someone higher up the hierarchy because of the changes to:
> 
> ```
> pkg/features/kube_features.go
> ```

FYI I am a feature approver, this should not require a top-level approver.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (ehashman) [inline pkg/kubelet/cm/container_manager_linux.go:335]: We probably want this to default to unset and only set if the feature gate is on, rather than the other way around?

Comment (fromanirh) [reply pkg/kubelet/cm/container_manager_linux.go:335]: good point, let me fix it

**Comment by fromanirh** [comment]
> > Although, this is mostly a kubelet-only change, it still requires approval from someone higher up the hierarchy because of the changes to:
> > ```
> > pkg/features/kube_features.go
> > ```
> 
> FYI I am a feature approver, this should not require a top-level approver.

Great :) so IIUC this (+ https://github.com/kubernetes/kubernetes/pull/102055#issuecomment-857894812) means than your approval + @derekwaynecarr / @dchen1107 's should enough to have this PR merged, right?
I updated the PR as per your suggestion and the lanes are green again.

**Comment by klueska** [comment]
> Great :) so IIUC this (+ [#102055 (comment)](https://github.com/kubernetes/kubernetes/pull/102055#issuecomment-857894812)) means than your approval + @derekwaynecarr / @dchen1107 's should enough to have this PR merged, right?


Hers plus mine should be enough.

**Comment by ranchothu** [comment]
Great, i will use this option.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (klueska) [inline pkg/kubelet/cm/container_manager_linux.go:336]: Setting this to `nil` here is unnecessary. It's initial value in the declaration is nil.

Also -- I think we should only print this warning if `nodeConfig.ExperimentalCPUManagerPolicyOptions` has some values set in it. Otherwise, the warning has no real meaning.

**Review by klueska — CHANGES_REQUESTED** [review]
A couple of small comments. Otherwise looks good to me.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (fromanirh) [reply pkg/kubelet/cm/container_manager_linux.go:336]: ack to both, thanks!

**Comment by klueska** [comment]
Looks like I was wrong about:
https://github.com/kubernetes/kubernetes/pull/102055#issuecomment-862179415

Inheritance is turned off in:
```
pkg/kubelet/apis/config/OWNERS
staging/src/k8s.io/kubelet/config/OWNERS
```
So we need an approval from someone in `api-approvers`, i.e. https://github.com/kubernetes/kubernetes/blob/master/OWNERS_ALIASES#L338

**Comment by fromanirh** [comment]
> Looks like I was wrong about:
> [#102055 (comment)](https://github.com/kubernetes/kubernetes/pull/102055#issuecomment-862179415)
> 
> Inheritance is turned off in:
> 
> ```
> pkg/kubelet/apis/config/OWNERS
> staging/src/k8s.io/kubelet/config/OWNERS
> ```
> 
> So we need an approval from someone in `api-approvers`, i.e. https://github.com/kubernetes/kubernetes/blob/master/OWNERS_ALIASES#L338

Thanks for checking! @thockin FYI

**Comment by fromanirh** [comment]
rebased and repushed to handle the conflict

**Comment by klueska** [comment]
/lgtm
/approve

@kubernetes/api-reviewers can we get a review from one of you?

File: pkg/kubelet/cm/container_manager_linux.go
Comment (liggitt) [inline pkg/kubelet/cm/container_manager_linux.go:336]: shouldn't setting values in the config for a disabled feature be caught at kubelet instantiation time and prevented?

Comment (liggitt) [reply pkg/kubelet/cm/container_manager_linux.go:336]: field doc just references the `CPUManager` feature, not CPUManagerPolicyOptions

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [inline pkg/kubelet/apis/config/types.go:226]: shouldn't the provided options be validated by the selected policy, and the "none" policy fail that validation if options are provided?

File: pkg/kubelet/cm/container_manager_linux.go
Comment (fromanirh) [reply pkg/kubelet/cm/container_manager_linux.go:336]: field doc: my fault, fixing.
re: kubelet instantation time: I can move the check earlier indeed. On it.

File: pkg/kubelet/apis/config/types.go
Comment (fromanirh) [reply pkg/kubelet/apis/config/types.go:226]: This is a good point. I made the `none` policy silently ignore everything because in the current incarnation it always succeed.
IMO giving options to the `none` policy is not wrong, it's just useless, hence the policy itself should ignore the options. But we were burned recently by a case on which we just logged and kept going, so I'm leaning towards making the policy fail if any options are given.

Comment (liggitt) [reply pkg/kubelet/apis/config/types.go]: > giving options to the none policy is not wrong, it's just useless

It seems like if we encounter that, the person configuring the kubelet made a mistake (typoed the cpuManager config field or something)... I think we should fail fast/loudly if we encounter configurations that make no sense in new fields we add

File: cmd/kubelet/app/options/options.go
Comment (fromanirh) [reply cmd/kubelet/app/options/options.go:494]: `NewMapStringStringNoSplit` seems the right thing here.

File: pkg/kubelet/apis/config/types.go
Comment (fromanirh) [reply pkg/kubelet/apis/config/types.go]: I think I addressed both these comments, mentioning both feature gates (reflecting what the code expects)

File: pkg/kubelet/apis/config/types.go
Comment (fromanirh) [reply pkg/kubelet/apis/config/types.go]: Yes, I think this approach is better and I changed the code accordingly.

File: cmd/kubelet/app/options/options.go
Comment (fromanirh) [reply cmd/kubelet/app/options/options.go:494]: To elaborate a bit: I don't see values including `=` or `,` on the horizon, but since this is already handled by the utilities and it seems more future proof I'm leaning towards using `NewMapStringStringNoSplit`. @klueska any preference from your side?

File: pkg/kubelet/apis/config/types.go
Comment (fromanirh) [reply pkg/kubelet/apis/config/types.go]: it now looks like: https://github.com/kubernetes/kubernetes/pull/101432/commits/2108e23c90e6f0f020ed5c1d8a970fe025f3a84d#diff-c11864857fa45d5c2d29c3a85d753ccd2db31cca1c7fa046b7144e71f6e59029R152 we will see during the review of PR 101432 if we want to move this guard inside the `none` policy proper or if can be kept here.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go
Comment (ehashman) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:146]: This doesn't appear to be used anywhere?

**Review by ehashman — COMMENTED** [review]
From pkg/features standpoint, this PR looks okay, but I would expect to see some additional unit test cases for the new kubelet config.

**Comment by klueska** [comment]
@ehashman its used in the PR it was split from https://github.com/kubernetes/kubernetes/pull/101432

The idea being to get the base infra in place and then have a follow-up PR with the first actual set of policy options in place.

**Comment by ehashman** [comment]
> @ehashman its used in the PR it was split from #101432
> 
> The idea being to get the base infra in place and then have a follow-up PR with the first actual set of policy options in place.

Sure, I just don't think there's any point in adding an argument and a bunch of nils to test signatures, we may as well make the change when the value is actually used.

**Comment by fromanirh** [comment]
> > @ehashman its used in the PR it was split from #101432
> > The idea being to get the base infra in place and then have a follow-up PR with the first actual set of policy options in place.
> 
> Sure, I just don't think there's any point in adding an argument and a bunch of nils to test signatures, we may as well make the change when the value is actually used.

point taken :) I guess we will not split next time.

**Comment by liggitt** [comment]
API change and mechanics (and the per-policy options validation in the next PR) lgtm.

Once functional PR is ready, this can be merged along with it.

**Comment by liggitt** [comment]
/lgtm

from an API mechanics perspective.

**Comment by fromanirh** [comment]
thanks @liggitt for your reviews!

**Comment by fromanirh** [comment]
rebased to address the conflicts with no code changes

**Comment by fromanirh** [comment]
rebased to fix the conflict.

I don't mind doing rebases to handle conflicts, but as we approach the code freeze, they are becoming more and more frequent. 

To minimize the _overall_ churn we should probably merge this PR - the followup PR consuming the changes here (https://github.com/kubernetes/kubernetes/pull/101432) is constantly up to date and got quite some review rounds already.

**Comment by liggitt** [comment]
> To minimize the _overall_ churn we should probably merge this PR

We avoid merging API before functional changes that use that API. One without the other doesn't accomplish much, and if the functional PR slips or has to be pulled out for some reason, having it paired with the API change has been helpful in thepast.

**Comment by fromanirh** [comment]
> > To minimize the _overall_ churn we should probably merge this PR
> 
> We avoid merging API before functional changes that use that API. One without the other doesn't accomplish much, and if the functional PR slips or has to be pulled out for some reason, having it paired with the API change has been helpful in the past.

Thanks for clarifying. Works for me.

**Comment by fromanirh** [comment]
@liggitt just a followup question, since https://github.com/kubernetes/kubernetes/pull/101432 builds on top of this PR - and includes all the commits in this PR - IIUC just merging https://github.com/kubernetes/kubernetes/pull/101432 once it gets all the needed acks seems the best thing, right?

**Comment by liggitt** [comment]
> @liggitt just a followup question, since #101432 builds on top of this PR - and includes all the commits in this PR - IIUC just merging #101432 once it gets all the needed acks seems the best thing, right?

yeah, that's fine as well.. can reference the lgtm here as API approval

**Comment by fromanirh** [comment]
> > @liggitt just a followup question, since #101432 builds on top of this PR - and includes all the commits in this PR - IIUC just merging #101432 once it gets all the needed acks seems the best thing, right?
> 
> yeah, that's fine as well.. can reference the lgtm here as API approval

Hi! just asking! Could you please review the API changes in https://github.com/kubernetes/kubernetes/pull/101432 so I can close this PR? 
101432 builds on top of these commits, just rebased more recently. Should be very quick.

**Comment by liggitt** [comment]
reviewed, verified API bits are identical in that PR... feel free to close this and I can tag that one once the non-API changes have review

**Comment by fromanirh** [comment]
> reviewed, verified API bits are identical in that PR... feel free to close this and I can tag that one once the non-API changes have review

many thanks! closing.

**Pull request closed without merging**
